### PR TITLE
Adding a new config parameter for the matching engine

### PR DESCRIPTION
**Pull request by aryx**
This is the first step towards enabling/disabling individual
features on a per-rule or per-pattern basis, e.g., disabling
constant propagation for some rules.

test plan:
make test

### CODE REVIEW

File: semgrep-core/Matching/Generic_vs_generic.ml [hunk 1]
```diff
@@ -598,11 +599,11 @@ and m_expr a b =
    * const a = "foo"; ... a == "foo" would be catched by $X == $X.
   *)
   | A.L(a1), b1 ->
-      (match Normalize_generic.constant_propagation_and_evaluate_literal b1 with
-       | Some b1 ->
-           m_literal_constness a1 b1
-       | None -> fail ()
-      )
+      if_config (fun x -> x.Config.constant_propagation)
```
Comment (aryx) [inline semgrep-core/Matching/Generic_vs_generic.ml:602]: @IagoAbal we can now enable/disable if we want.

Comment (aryx) [reply semgrep-core/Matching/Generic_vs_generic.ml:602]: Remains to hook this up in the rule format to let user enable/disable equivalences/features.

Comment (IagoAbal) [reply semgrep-core/Matching/Generic_vs_generic.ml:602]: Great!

**Review by IagoAbal — APPROVED** [review]
Looks great to me.

**Comment by minusworld** [comment]
@aryx It looks like this introduces a Config object. How would I enable/disable constant propagation using `semgrep-core`?

**Comment by aryx** [comment]
we need to add a field in the rule to allow enable/disable individual features. This will be easier once we do the "big refactoring" in evaluation.py where we just pass the whole rule file to semgrep-core.

**Pull request merged**